File: lev1/짝수와 홀수.py
"""
문제 설명
정수 num이 짝수일 경우 Even을 반환하고 홀수인 경우 Odd를 반환하는 함수, solution을 완성해주세요.

제한 조건
num은 int 범위의 정수입니다.
0은 짝수입니다.

입출력 예
num	return
3	Odd
4	Even
"""


def solution(num):
    # 0: 00:00.048994
    # return "Odd" if num % 2 else "Even"

    # code refactoring
    # 0:00:00.060079
    # % 최적화가 잘되어 있어서 비트연산이 더 느린 것으로 추측 (구현체 자체가 더 빨라서)
    return 'Odd' if num & 1 else 'Even'


print(solution(3))
print(solution(4))
print(solution(44))
print(solution(5))

from datetime import datetime
s = datetime.now()
for _ in range(100000):
    solution(3)
    solution(4)
    solution(44)
    solution(5)
print(datetime.now() - s)

File: lev1/짝수와홀수.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

char* solution(int num) {
    // // 0.017163
    // char* answer = (char*)malloc(4);
    // if (num % 2 == 0)
    //     answer = "Even";
    // else
    //     answer = "Odd";

    // code refactoring - 0.014269
    int d = num%2;
    char* answer = (char*)malloc(sizeof(char)*(5-d));
    answer = d ? "Odd" : "Even";
    return answer;
    
};

int main(){
    clock_t tic = clock();
    for (int i=0; i<100000; i++){
        solution(3);
        solution(2147483647);
        solution(2147483646);
    }
    clock_t toc = clock();
    printf("running time: %f\n", (double)(toc-tic) / CLOCKS_PER_SEC);
    return 0;
}


